h5tovtk.1
.\" Copyright (c) 2002 Massachusetts Institute of Technology
.\" 
.\" Permission is hereby granted, free of charge, to any person obtaining
.\" a copy of this software and associated documentation files (the
.\" "Software"), to deal in the Software without restriction, including
.\" without limitation the rights to use, copy, modify, merge, publish,
.\" distribute, sublicense, and/or sell copies of the Software, and to
.\" permit persons to whom the Software is furnished to do so, subject to
.\" the following conditions:
.\" 
.\" The above copyright notice and this permission notice shall be
.\" included in all copies or substantial portions of the Software.
.\" 
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
.\" EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
.\" IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
.\" CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
.\" TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
.\" SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
.\"
.TH H5TOVTK 1 "March 9, 2002" "h5utils" "h5utils"
.SH NAME
h5tovtk \- convert datasets in HDF5 files to VTK format
.SH SYNOPSIS
.B h5tovtk
[\fIOPTION\fR]... [\fIHDF5FILE\fR]...
.SH DESCRIPTION
.PP
." Add any additional description here
h5tovtk is a program to generate VTK data files from multidimensional
datasets in HDF5 files.  VTK, the Visualization ToolKit, is an
open-source, freely available software system for 3D computer
graphics, image processing, and visualization.  VTK itself is a
programming library, but it is also the basis for a number of end-user
graphical visualization programs.

HDF5 is a free, portable binary format and supporting library developed
by the National Center for Supercomputing Applications at the University
of Illinois in Urbana-Champaign.  A single
.I h5
file can contain multiple datasets; by default,
.I h5tovtk
takes the first dataset, but this can be changed via the
.B -d
option, or by using the syntax \fIHDF5FILE:DATASET\fR.

1d/2d/3d datasets are converted into 3d VTK \"structured points\"
datasets.  Normally, a single scalar VTK dataset is output, but
vectors and fields can be output via the
.B -o
option below.

A typical invocation is of the form
\'h5tovtk foo.h5\', which will output a VTK data file foo.vtk
from the data in foo.h5.
.SH OPTIONS
.TP
.B -h
Display help on the command-line options and usage.
.TP
.B -V
Print the version number and copyright info for h5tovtk.
.TP
.B -v
Verbose output.
.TP
\fB\-o\fR \fIfile\fR
Save all the input datasets to a single VTK \fIfile\fR.  If there is
only one dataset, it is output to a VTK scalar dataset; if there are
three datasets, they are output as a VTK vector dataset; all other
numbers of datasets are combined into a VTK field dataset.

Otherwise, the default behavior is to save each dataset to a separate
VTK file, with the .h5 suffix of the input filename replaced by .vtk
in the output filename.

Only three-dimensional datasets may be written to the VTK file.  If
you have a four (or more) dimensional data set, then you must take a
three-dimensional "slice" of the multi-dimensional data.  To do this,
you specify coordinates in one (or more) slice dimension(s), via the
.B -xyzt
options.
.SH OPTIONS
.TP
\fB\-1\fR, \fB\-2\fR, \fB\-4\fR
Use 1 , 2, or 4 bytes to store each data point in the
output file.  Fewer bytes require less storage and memory,
but will decrease the resolution in the values.
.B -1
will break up the data values into one of 256 possible values (on a
linear scale from the minimum to the maximum value in your data),
.B -2
will allow 65536 possible values, and
.B -4
(the default) will use 4-byte floating-point numbers for an "exact"
representation.
.TP
.B -a
Output in ASCII format; otherwise, VTK's more compact, but less
readable and somewhat less portable binary format is used.
.TP
.B -n
For binary output (see
.B -a
above), by default the data is written in bigendian byte order, which
is normally the order that VTK expects.  However, some external tools
and a few VTK classes use the native byte ordering instead (which may
not be bigendian), and the
.B -n
option causes
.I h5tovtk
to output binary data in the native ordering.
.TP
\fB\-m\fR \fImin\fR, \fB\-M\fR \fImax\fR
When
.B -1
or
.B -2
are used, the input data are converted to a linear integer scale.
Normally, the bottom and top of this scale correspond to the
minimum and maximum values in the data.  Using the
.B -m
and
.B -M
options, you
can make the bottom and top of the scale correspond to
.I min
and
.I max
instead, respectively.  Data values below or above this range will be
treated as if they were
.I min
or
.I max
respectively.  See also the
.B -Z
option.
.TP
.B -Z
For
.B -1
or
.B -2
output, center the linear integer scale on the value zero in the data.
.TP
.B -r
Invert the output values (map the minimum to the maximum and vice versa).
.TP
\fB\-x\fR \fIix\fR, \fB\-y\fR \fIiy\fR, \fB\-z\fR \fIiz\fR, \fB\-t\fR \fIit\fR
This tells
.I h5tovtk
to use a particular slice of a multi-dimensional dataset.  e.g.
.B -x
uses the subset (with one less dimension) at an x index of
.I ix
(where the indices run from zero to one less than the maximum index in
that direction).  Here, x/y/z correspond to the first/second/third
dimensions of the HDF5 dataset. The \fB\-t\fR option specifies a slice
in the last dimension, whichever that might be.  See also the
.B -0
option to shift the origin of the x/y/z slice coordinates to the
dataset center.
.TP
.B -0
Shift the origin of the x/y/z slice coordinates to the dataset center,
so that e.g. -0 -x 0 (or more compactly -0x0) returns the central x
plane of the dataset instead of the edge x plane.  (\fB\-t\fR
coordinates are not affected.)
.TP
\fB\-d\fR \fIname\fR
Use dataset
.I name
from the input files; otherwise, the first dataset from each file is used.
Alternatively, use the syntax \fIHDF5FILE:DATASET\fR, which allows you
to specify a different dataset for each file.
You can use the
.I h5ls
command (included with hdf5) to find the names of datasets within a file.
.SH BUGS
Send bug reports to S. G. Johnson, stevenj@alum.mit.edu.
.SH AUTHORS
Written by Steven G. Johnson.  Copyright (c) 2005 by the Massachusetts
Institute of Technology.


NEWS
h5utils 1.10 (9/2/05)

  * Added h4fromh5 utility.  (NCSA seems to be no longer shipping
    the HDF5/HDF4 conversion tools with the latest HDF5 release.)

  * Added dkbluered color map, which is similar to bluered but
    uses a somewhat wider range of colors.

h5utils 1.9.1 (8/5/04)

  * Fix h5topng compilation failure with some non-C99 compilers;
    thanks to Maarten van Reeuwijk for the bug report.

h5utils 1.9 (7/12/04)

  * Added new h5math utility, which creates and combines HDF5 datasets
    using a user-specified mathematical expression.  (Requires GNU
    libmatheval to be installed.)

  * h5topng: Matlab-like start:end or start:step:end notation for
    slice indices, to allow a whole sequence of slices to be output as
    a sequence of PNG images.

  * h5topng: if contour/overlay dataset does not have same dimensions
    as output data, it is periodically "tiled" over the output.

h5utils 1.8 (5/22/04)

  * New -A and -a options for h5topng to allow translucent overlays
    from one file onto another (an alternative or complement to the
    -C contour-overlay option).

  * h5topng uses 24-bit direct color by default (use -8 option for
    old 8-bit behavior).

  * h5topng uses columns/rows for x/y by default (use -T to swap),
    the opposite of the old behavior.

  * There is no default -z 0 slice dimension in h5topng/h5totxt any more.
    You must specify a slice for 3+ dimensional data in h5topng.
    h5totxt dumps the whole data file by default unless one or more
    slices are specified.

  * Support specifying multiple slice dimensions for 4+ dimensional
    datasets, with new "-t" option to indicate final dimension.

  * Slices are also supported now in h5tovtk and h5tov5d.

  * New -. option in h5totxt to specify number of significant digits; 
    output 16 digits by default instead of 6, previously.

h5utils 1.7.1 (3/16/02)

  * Fixed array overrun in h5topng that caused a floating-point exception
    on Alphas; thanks to Marin Soljacic for the bug report.

h5utils 1.7 (3/9/02)

  * h5topng now supports multiple, user-definable color tables, a number
    of which are provided.  INCOMPATIBLE CHANGE: the -c option now has
    the syntax: -c <colortable>.  The old behavior corresponds to the
    included "bluered" colortable, invoked via: -c bluered.

  * New -R option for h5topng to use a consistent color scale for
    all specified files.

  * New -0 option for h5topng and h5totxt that shifts the origin of the
    slice coordinates to the dataset center.

  * Added h5tovtk program to output VTK (Visualization ToolKit) data files.

  * Support -T (transpose dimensions) option in h5tov5d.

  * Fixed bug in h5topng that caused extra rows/columns of garbage pixels
    to be written at the edges of images when scaling was used.

  * When compiling the h5read Octave plugin, respect the CPPFLAGS and LDFLAGS 
    environment variables.  Thanks to Max Colice for the bug report.

  * Fixed problem when --without-h5tov5d and --without-h5fromh4 are used. 
    Thanks to Nikola Ivanov Nikolov for the fix.

h5utils 1.6 (1/17/01)

  * Don't build h5fromh4 if the superior h4toh5 tool (from HDF5 1.4)
    is present.  Also added --{with,without}-h5fromh4 option to
    configure to force whether h5fromh4 is built.

h5utils 1.5.1 (12/9/00)

  * Support manually disabling Octave plugin support
    (configure --without-octave) in case of C++ problems.

  * Support Vis5d+ and Debian HDF header file locations.

h5utils 1.5 (7/9/00)

  * Added h5fromh4 program to convert HDF4 datasets to HDF5 format.

  * Added -S <s> option to h5topng as a shortcut for -X <s> -Y <s>.

h5utils 1.4 (5/28/00)

  * Added h5fromtxt program to convert text input to an HDF5 dataset.

h5utils 1.3.4 (1/31/00)

  * Improved -C contour plotting in h5topng.

  * Fix in h5topng man page (thanks to Christoph Becher).

h5utils 1.3.3 (1/30/00)

  * Bug fix in h5topng (would sometimes output solid black images).
    Thanks to Karl Koch for the bug report.

h5utils 1.3.2 (1/28/00)

  * Added h5topng -Z option to center color scale on zero.

  * Now support h5topng -C filename:dataset.

h5utils 1.3.1 (1/27/00)

  * Bug fixes in dataset name-handling, especially when using h5tov5d to
    join multiple datasets into one output file.

h5utils 1.3 (1/21/00)

  * You can now specify individual datasets within a file by using
    <filename>:<dataset> instead of just <filename> with h5topng,
    h5totxt, and h5tov5d.

h5utils 1.2.3: (1/20/00)

  * Fixed minor bug in 'h5totxt -h'.

h5utils 1.2.2: (1/12/00)

  * Makefile now includes CPPFLAGS in the compiler flags, making it
    easier to use header files in non-standard locations.  (CPPFLAGS
    is the proper place to put -I flags for the configure script.)

h5utils 1.2.1:  (1/1/00)

  * Modified -o option of h5tov5d to join datasets into a single Vis5d file.

h5utils 1.2:  (12/31/99)

  * Added h5tov5d program for converting to Vis5d format.

  * Improved installation; h5read.oct now goes into the sitewide Octave
    plugins directory, and things work correctly when only some of the
    utilities are compiled.

h5utils 1.1:  (12/6/99)

  * Added h5totxt program for exporting 2d slices of HDF5 files to text
    suitable for importing into a spreadsheet.

h5utils 1.0:  (11/22/99)

  * Initial release.


h5tovtk.c
/* Copyright (c) 2005 Massachusetts Institute of Technology
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 * 
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <ctype.h>

#include <unistd.h>

#ifdef HAVE_NETINET_IN_H
#  include <netinet/in.h>
#endif

#include "config.h"
#include "arrayh5.h"
#include "copyright.h"
#include "h5utils.h"

#define CHECK(cond, msg) { if (!(cond)) { fprintf(stderr, "h5tovtk error: %s\n", msg); exit(EXIT_FAILURE); } }

void usage(FILE *f)
{
     fprintf(f, "Usage: h5tovtk [options] [<filenames>]\n"
	     "Options:\n"
	     "         -h : this help message\n"
             "         -V : print version number and copyright\n"
	     "         -v : verbose output\n"
	     "  -o <file> : output datasets from all input files to <file>;\n"
	     "              combines 3 datasets to a vector, and 2 or 4+ to a field\n"
	     "         -4 : 4-byte floating-point binary output (default)\n"
	     "         -a : ASCII floating-point output\n"
	     "         -1 : 1-byte (0-255) binary output\n"
	     "         -2 : 2-byte (0-65535) binary output\n"
	     "         -n : don't convert binary output to bigendian\n"
	     "   -m <min> : set bottom of scale for 1/2 byte encoding\n"
	     "   -M <max> : set top of scale for 1/2 byte encoding\n"
	     "         -Z : center scale at zero for 1/2 byte encoding\n"
	     "         -r : invert scale & data values\n"
	     "    -x <ix> : take x=<ix> slice of data\n"
	     "    -y <iy> : take y=<iy> slice of data\n"
	     "    -z <iz> : take z=<iz> slice of data\n"
	     "    -t <it> : take t=<it> slice of data's last dimension\n"
	     "         -0 : use dataset center as origin for -x/-y/-z\n"
	     "  -d <name> : use dataset <name> in the input files (default: first dataset)\n"
	     "              -- you can also specify a dataset via <filename>:<name>\n"
	  );
}

static void whitespace_to_underscores(char *s)
{
     while (*s) {
	  if (isspace(*s))
	       *s = '-';
	  ++s;
     }
}

static const char vtk_datatype[][20] = { 
     "float", "unsigned_char", "unsigned_short", "none", "float"
};

static void write_vtk_header(FILE *f, int is_binary,
			     int nx, int ny, int nz,
			     double ox, double oy, double oz,
			     double sx, double sy, double sz)
{
     fprintf(f, 
	     "# vtk DataFile Version 2.0\n"
	     "Generated by h5tovtk.\n"
	     "%s\n"
	     "DATASET STRUCTURED_POINTS\n"
	     "DIMENSIONS %d %d %d\n"
	     "ORIGIN %g %g %g\n"
	     "SPACING %g %g %g\n",
	     is_binary ? "BINARY" : "ASCII",
	     nz, ny, nx,
	     oz, oy, ox, sz, sy, sx);
}

static void write_vtk_value(FILE *f, double v, int store_bytes, int fix_bytes,
			    double min, double max, int invert)
{
     if (invert)
	  v = max - (v - min);
     switch (store_bytes) {
	 case 0:
	      fprintf(f, "%g ", v);
	      break;
	 case 1:
	 {
	      unsigned char c;
	      c = floor((v - min) * 255.0 / (max - min) + 0.5);
	      fwrite(&c, 1, 1, f);
	      break;
	 }
	 case 2:
	 {
	      unsigned short i;
	      i = floor((v - min) * 65535.0 / (max - min) + 0.5);
	      if (fix_bytes) {
#if defined(HAVE_HTONS)
		   i = htons(i);
#elif ! defined(WORDS_BIGENDIAN)
		   unsigned char swap, *bytes;
		   bytes = (unsigned char *) &i;
		   swap = bytes[0]; bytes[0] = bytes[1]; bytes[1] = swap;
#endif
	      }
	      fwrite(&i, 2, 1, f);
	      break;
	 }
	 case 4:
	 {
	      float fv = v;
	      if (fix_bytes) {
#if defined(HAVE_HTONL) && (SIZEOF_FLOAT == SIZEOF_UNSIGNED_LONG)
		   unsigned long *i = (unsigned long *) &fv;
		   *i = htonl(*i);
#elif ! defined(WORDS_BIGENDIAN)
		   unsigned char swap, *bytes;
		   bytes = (unsigned char *) &fv;
		   swap = bytes[0]; bytes[0] = bytes[3]; bytes[3] = swap;
		   swap = bytes[1]; bytes[1] = bytes[2]; bytes[2] = swap;
#endif
	      }
	      fwrite(&fv, 4, 1, f);
	      break;
	 }
     }
}

int main(int argc, char **argv)
{
     arrayh5 *a = NULL;
     char *vtk_fname = NULL, *data_name = NULL;
     extern char *optarg;
     extern int optind;
     double ox = 0, oy = 0, oz = 0, sx = 1, sy = 1, sz = 1;
     int c, ifile;
     int zero_center = 0;
     int invert = 0;
     double min = 0, max = 0;
     int min_set = 0, max_set = 0;
     int verbose = 0, combine = 0;
     int slicedim[4] = {NO_SLICE_DIM,NO_SLICE_DIM,NO_SLICE_DIM,NO_SLICE_DIM};
     int islice[4], center_slice[4] = {0,0,0,0};
     int nx = 0, ny = 0, nz = 0, na;
     int store_bytes = 4, fix_byte_order = 1;

     while ((c = getopt(argc, argv, "ho:d:vV124mMZranx:y:z:t:0")) != -1)
	  switch (c) {
	      case 'h':
		   usage(stdout);
		   return EXIT_SUCCESS;
	      case 'V':
		   printf("h5tovtk " PACKAGE_VERSION " by Steven G. Johnson\n" 
			  COPYRIGHT);
		   return EXIT_SUCCESS;
	      case 'v':
		   verbose = 1;
		   break;
	      case 'x':
		   islice[0] = atoi(optarg);
		   slicedim[0] = 0;
		   break;
	      case 'y':
		   islice[1] = atoi(optarg);
		   slicedim[1] = 1;
		   break;
	      case 'z':
		   islice[2] = atoi(optarg);
		   slicedim[2] = 2;
		   break;
	      case 't':
		   islice[3] = atoi(optarg);
		   slicedim[3] = LAST_SLICE_DIM;
		   break;
              case '0':
                   center_slice[0] = center_slice[1] = center_slice[2] = 1;
                   break;
	      case 'n':
		   fix_byte_order = 0;
		   break;
	      case 'a':
		   store_bytes = 0; /* ascii */
		   break;
	      case '1':
		   store_bytes = 1;
		   break;
	      case '2':
		   store_bytes = 2;
		   break;
	      case '4':
		   store_bytes = 4;
		   break;
	      case 'm':
		   min = atof(optarg);
		   min_set = 1;
		   break;
	      case 'M':
		   max = atof(optarg);
		   max_set = 1;
		   break;
	      case 'Z':
		   zero_center = 1;
		   break;
	      case 'r':
		   invert = 1;
		   break;
	      case 'o':
		   vtk_fname = my_strdup(optarg);
		   combine = 1;
		   break;
	      case 'd':
		   data_name = my_strdup(optarg);
		   break;		   
	      default:
		   fprintf(stderr, "Invalid argument -%c\n", c);
		   usage(stderr);
		   return EXIT_FAILURE;
	  }
     if (optind == argc) {  /* no parameters left */
	  usage(stderr);
	  return EXIT_FAILURE;
     }

     CHECK(store_bytes != 4 || sizeof(float) == 4, 
	   "'float' is wrong size for -4");
     CHECK(store_bytes != 2 || sizeof(unsigned short) == 4, 
	   "'short' is wrong size for -2");
     
     a = (arrayh5*) malloc(sizeof(arrayh5) * (na = argc - optind));
     CHECK(a, "out of memory");

     combine = combine && (na > 1);

     for (ifile = optind; ifile < argc; ++ifile) {
          char *dname, *found_dname, *h5_fname;
	  int err, ia = ifile - optind;
          h5_fname = split_fname(argv[ifile], &dname);
          if (!dname[0])
               dname = data_name;

	  err = arrayh5_read(&a[ia], h5_fname, dname, &found_dname,
			     4, slicedim, islice, center_slice);
	  CHECK(!err, arrayh5_read_strerror[err]);
	  CHECK(a[ia].rank >= 1, "data must have at least one dimension");
	  CHECK(a[ia].rank <= 3, "data can have at most 3 dimensions (try taking a slice");
	  
	  CHECK(!combine || !ia || arrayh5_conformant(a[ia], a[0]),
		"all arrays must be conformant to combine them");
	  
	  if (!vtk_fname)
	       vtk_fname = replace_suffix(h5_fname, ".h5", ".vtk");

	  {
	       double a_min, a_max;
	       arrayh5_getrange(a[ia], &a_min, &a_max);
	       if (verbose)
		    printf("data in %s ranges from %g to %g.\n", 
			   h5_fname, a_min, a_max);
	       if (!min_set)
		    min = (!combine || !ia || a_min < min) ? a_min : min;
	       if (!max_set)
		    max = (!combine || !ia || a_max > max) ? a_max : max;
	       if (min > max) {
		    invert = !invert;
		    a_min = min;
		    min = max;
		    max = a_min;
	       }
	       if (zero_center) {
		    max = fabs(max) > fabs(min) ? fabs(max) : fabs(min);
		    min = -max;
	       }
	  }
	  
	  nx = a[ia].dims[0];
	  ny = a[ia].rank < 2 ? 1 : a[ia].dims[1];
	  nz = a[ia].rank < 3 ? 1 : a[ia].dims[2];
	  
	  if (!combine) {
	       FILE *f;
	       int i, N = nx * ny * nz;

	       if (verbose)
		    printf("writing \"%s\" from %dx%dx%d input data.\n",
			   vtk_fname, nx, ny, nz);
	       
	       if (strcmp(vtk_fname, "-")) {
		    f = fopen(vtk_fname, "w");
		    CHECK(f, "error creating file");
	       }
	       else
		    f = stdout;

	       write_vtk_header(f, store_bytes, 
				nx, ny, nz, ox, oy, oz, sx, sy, sz);
	       whitespace_to_underscores(found_dname);
	       fprintf(f, "POINT_DATA %d\n"
		       "SCALARS %s %s 1\n"
		       "LOOKUP_TABLE default\n",
		       N, found_dname, vtk_datatype[store_bytes]);
	       
	       for (i = 0; i < N; ++i)
		    write_vtk_value(f, a[ia].data[i], store_bytes,
				    fix_byte_order, min, max, invert);
	  
	       if (f != stdout)
		    fclose(f);
	       arrayh5_destroy(a[ia]);
	       free(vtk_fname); vtk_fname = NULL;
	  }
	  free(found_dname);
	  free(h5_fname);
     }

     if (combine) {
	  FILE *f;
	  int i, N = nx * ny * nz;

	  if (verbose)
	       printf("writing \"%s\" from %dx%dx%d input data.\n",
		      vtk_fname, nx, ny, nz);
	  
	  if (strcmp(vtk_fname, "-")) {
	       f = fopen(vtk_fname, "w");
	       CHECK(f, "error creating file");
	  }
	  else
	       f = stdout;
	  
	  write_vtk_header(f, store_bytes, 
			   nx, ny, nz, ox, oy, oz, sx, sy, sz);
	  fprintf(f, "POINT_DATA %d\n", N);
	  switch (na) {
	      case 1:
		   fprintf(f, "SCALARS scalars %s 1\nLOOKUP_TABLE default\n", 
			   vtk_datatype[store_bytes]);
		   break;
	      case 3:
		   fprintf(f, "VECTORS vectors %s\n", 
			   vtk_datatype[store_bytes]);
		   break;
	      default:
		   fprintf(f, "FIELD fields 1\narray %d %d %s\n", 
			   na, N, vtk_datatype[store_bytes]);
	  }
	  
	  for (i = 0; i < N; ++i) {
	       int ia;
	       for (ia = 0; ia < na; ++ia)
		    write_vtk_value(f, a[ia].data[i], store_bytes, 
				    fix_byte_order, min, max, invert);
	  }
	  if (f != stdout)
	       fclose(f);
	  {
	       int ia;
	       for (ia = 0; ia < na; ++ia)
		    arrayh5_destroy(a[ia]);
	  }
     }

     free(a);

     if (data_name)
	  free(data_name);

     return EXIT_SUCCESS;
}
